File: web/src/views/system/schema/parseOpenSpgSchema.ts
/**
 * Parser for OpenSPG .schema DSL (namespace + EntityType blocks).
 */

export interface ParsedSchemaProperty {
	nameEn: string;
	nameZh: string;
	type: string;
	index?: string;
	constraints: string[];
}

export interface ParsedSchemaRelation {
	nameEn: string;
	nameZh: string;
	targetType: string;
}

export interface ParsedSchemaEntity {
	nameEn: string;
	nameZh: string;
	desc?: string;
	properties: ParsedSchemaProperty[];
	relations: ParsedSchemaRelation[];
}

export interface ParsedOpenSpgSchema {
	namespace: string;
	entities: ParsedSchemaEntity[];
}

export interface ParsedSchemaStats {
	namespace: string;
	entities: number;
	relations: number;
	properties: number;
	constraint: number;
	indexCount: number;
}

const ENTITY_HEADER_RE = /^([A-Za-z_]\w*)\(([^)]*)\):\s*EntityType\s*$/;
const NAMED_FIELD_RE = /^(\w+)\(([^)]*)\):\s*(\S+)\s*$/;
const NAMESPACE_RE = /^namespace\s+([A-Za-z_]\w*)\s*$/;

function stripComments(line: string): string {
	const idx = line.indexOf('#');
	return idx >= 0 ? line.slice(0, idx) : line;
}

function leadingIndent(line: string): number {
	const m = line.match(/^(\s*)/);
	return m ? m[1].replace(/\t/g, '  ').length : 0;
}

export function parseOpenSpgSchema(raw: string): ParsedOpenSpgSchema {
	const text = raw.replace(/^\uFEFF/, '').replace(/\r\n/g, '\n');
	const lines = text.split('\n');

	let namespace: string | null = null;
	const entities: ParsedSchemaEntity[] = [];
	let current: ParsedSchemaEntity | null = null;
	let section: 'none' | 'properties' | 'relations' = 'none';
	let lastProperty: ParsedSchemaProperty | null = null;
	let entitySectionIndent = 0;

	for (let i = 0; i < lines.length; i++) {
		const line = stripComments(lines[i]).trimEnd();
		const trimmed = line.trim();
		if (!trimmed) continue;

		const indent = leadingIndent(line);

		const nsMatch = trimmed.match(NAMESPACE_RE);
		if (nsMatch) {
			if (entities.length > 0 || current) {
				throw new SyntaxError(`Line ${i + 1}: 'namespace' must appear before entity definitions.`);
			}
			namespace = nsMatch[1];
			continue;
		}

		const entityMatch = trimmed.match(ENTITY_HEADER_RE);
		if (entityMatch && indent === 0) {
			current = {
				nameEn: entityMatch[1],
				nameZh: entityMatch[2].trim(),
				properties: [],
				relations: [],
			};
			entities.push(current);
			section = 'none';
			lastProperty = null;
			entitySectionIndent = 0;
			continue;
		}

		if (!current) {
			if (!namespace) {
				throw new SyntaxError(
					`Line ${i + 1}: Missing 'namespace' definition at the beginning of the file.`
				);
			}
			throw new SyntaxError(`Line ${i + 1}: Unexpected content before first EntityType.`);
		}

		if (/^desc:\s*(.+)$/.test(trimmed)) {
			current.desc = trimmed.replace(/^desc:\s*/, '').trim();
			continue;
		}

		if (/^properties:\s*$/.test(trimmed)) {
			section = 'properties';
			entitySectionIndent = indent;
			lastProperty = null;
			continue;
		}

		if (/^relations:\s*$/.test(trimmed)) {
			section = 'relations';
			entitySectionIndent = indent;
			lastProperty = null;
			continue;
		}

		const indexMatch = trimmed.match(/^index:\s*(.+)$/);
		if (indexMatch && lastProperty && section === 'properties') {
			lastProperty.index = indexMatch[1].trim();
			continue;
		}

		const constraintMatch = trimmed.match(/^constraint:\s*(.+)$/);
		if (constraintMatch && lastProperty && section === 'properties') {
			lastProperty.constraints = constraintMatch[1]
				.split(',')
				.map((c) => c.trim())
				.filter(Boolean);
			continue;
		}

		const fieldMatch = trimmed.match(NAMED_FIELD_RE);
		if (fieldMatch && indent > entitySectionIndent) {
			const [, nameEn, nameZh, type] = fieldMatch;
			if (section === 'properties') {
				const prop: ParsedSchemaProperty = {
					nameEn,
					nameZh: nameZh.trim(),
					type,
					constraints: [],
				};
				current.properties.push(prop);
				lastProperty = prop;
			} else if (section === 'relations') {
				current.relations.push({
					nameEn,
					nameZh: nameZh.trim(),
					targetType: type,
				});
				lastProperty = null;
			}
			continue;
		}
	}

	if (!namespace) {
		throw new SyntaxError("Missing 'namespace' definition at the beginning of the file.");
	}

	if (entities.length === 0) {
		throw new SyntaxError('No EntityType definitions found in schema file.');
	}

	return { namespace, entities };
}

export function getParsedSchemaStats(parsed: ParsedOpenSpgSchema): ParsedSchemaStats {
	let properties = 0;
	let constraint = 0;
	let indexCount = 0;
	let relations = 0;

	for (const entity of parsed.entities) {
		properties += entity.properties.length;
		relations += entity.relations.length;
		for (const prop of entity.properties) {
			if (prop.constraints.length > 0) constraint += 1;
			if (prop.index) indexCount += 1;
		}
	}

	return {
		namespace: parsed.namespace,
		entities: parsed.entities.length,
		relations,
		properties,
		constraint,
		indexCount,
	};
}

File: backend/schema_manager/parseOpenSpgSchema.py
"""Parser for OpenSPG .schema DSL — ported from the frontend parseOpenSpgSchema.ts."""

import re
from dataclasses import dataclass, field


@dataclass
class ParsedSchemaProperty:
    nameEn: str
    nameZh: str
    type: str
    index: str | None = None
    constraints: list[str] = field(default_factory=list)


@dataclass
class ParsedSchemaRelation:
    nameEn: str
    nameZh: str
    targetType: str


@dataclass
class ParsedSchemaEntity:
    nameEn: str
    nameZh: str
    desc: str | None = None
    properties: list[ParsedSchemaProperty] = field(default_factory=list)
    relations: list[ParsedSchemaRelation] = field(default_factory=list)


ENTITY_HEADER_RE = re.compile(r"^([A-Za-z_]\w*)\(([^)]*)\):\s*EntityType\s*$")
NAMED_FIELD_RE = re.compile(r"^(\w+)\(([^)]*)\):\s*(\S+)\s*$")
NAMESPACE_RE = re.compile(r"^namespace\s+([A-Za-z_]\w*)\s*$")


def _strip_comments(line: str) -> str:
    idx = line.find("#")
    return line[:idx] if idx >= 0 else line


def _leading_indent(line: str) -> int:
    return len(line) - len(line.lstrip())


def parseOpenSpgSchema(raw: str) -> dict:
    text = raw.lstrip("﻿").replace("\r\n", "\n")
    lines = text.split("\n")

    namespace = None
    entities: list[ParsedSchemaEntity] = []
    current: ParsedSchemaEntity | None = None
    section: str = "none"  # "none" | "properties" | "relations"
    last_property: ParsedSchemaProperty | None = None
    entity_section_indent = 0

    for i, raw_line in enumerate(lines):
        line = _strip_comments(raw_line).rstrip()
        trimmed = line.strip()
        if not trimmed:
            continue

        indent = _leading_indent(line)

        ns_match = NAMESPACE_RE.match(trimmed)
        if ns_match:
            if entities or current:
                raise SyntaxError(
                    f"Line {i + 1}: 'namespace' must appear before entity definitions."
                )
            namespace = ns_match.group(1)
            continue

        entity_match = ENTITY_HEADER_RE.match(trimmed)
        if entity_match and indent == 0:
            current = ParsedSchemaEntity(
                nameEn=entity_match.group(1),
                nameZh=entity_match.group(2).strip(),
            )
            entities.append(current)
            section = "none"
            last_property = None
            entity_section_indent = 0
            continue

        if current is None:
            if namespace is None:
                raise SyntaxError(
                    f"Line {i + 1}: Missing 'namespace' definition at the beginning of the file."
                )
            raise SyntaxError(f"Line {i + 1}: Unexpected content before first EntityType.")

        if trimmed.startswith("desc:"):
            current.desc = trimmed.removeprefix("desc:").strip()
            continue

        if trimmed == "properties:":
            section = "properties"
            entity_section_indent = indent
            last_property = None
            continue

        if trimmed == "relations:":
            section = "relations"
            entity_section_indent = indent
            last_property = None
            continue

        if trimmed.startswith("index:") and last_property and section == "properties":
            last_property.index = trimmed.removeprefix("index:").strip()
            continue

        if trimmed.startswith("constraint:") and last_property and section == "properties":
            last_property.constraints = [
                c.strip() for c in trimmed.removeprefix("constraint:").split(",") if c.strip()
            ]
            continue

        field_match = NAMED_FIELD_RE.match(trimmed)
        if field_match and indent > entity_section_indent:
            name_en, name_zh, type_ = field_match.groups()
            if section == "properties":
                prop = ParsedSchemaProperty(
                    nameEn=name_en,
                    nameZh=name_zh.strip(),
                    type=type_,
                )
                current.properties.append(prop)
                last_property = prop
            elif section == "relations":
                current.relations.append(
                    ParsedSchemaRelation(
                        nameEn=name_en,
                        nameZh=name_zh.strip(),
                        targetType=type_,
                    )
                )
                last_property = None
            continue

    if namespace is None:
        raise SyntaxError("Missing 'namespace' definition at the beginning of the file.")
    if not entities:
        raise SyntaxError("No EntityType definitions found in schema file.")

    return {
        "namespace": namespace,
        "entities": entities,
    }


def getParsedSchemaStats(parsed: dict) -> dict:
    entities = parsed["entities"]
    properties = 0
    constraint = 0
    index_count = 0
    relations = 0

    for entity in entities:
        properties += len(entity.properties)
        relations += len(entity.relations)
        for prop in entity.properties:
            if prop.constraints:
                constraint += 1
            if prop.index:
                index_count += 1

    return {
        "namespace": parsed["namespace"],
        "entities": len(entities),
        "relations": relations,
        "properties": properties,
        "constraint": constraint,
        "indexCount": index_count,
    }
